File: trunk/xchat-gnome/plugins/url_scraper/UrlScraper.py
''' UrlScraper.py

xchat-gnome script for pulling URLs out of channels
and displaying them in a separate window.
'''
# Copyright (C) 2004 W. Evan Sheehan <evan@navi.cx>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import xchat, string, re, gconf
import pygtk
import gtk
import gobject

__module_name__ = 'URL Scraper'
__module_version__ = '0.1-pre'
__module_description__ = 'Collect URLs said and display them in a separate window'


def close( window, event=None, user_data=None ):
    ''' Unload the module when you close the window. '''
    # FIXME: In the future this might just hide the window if we can add a
    #        item to the main window that would allow us to show the window,
    #        in which case we wouldn't unload the plugin here.
    xchat.command( 'py unload '+__module_name__ )

def grabbedURL( nick, match ):
    ''' Handler for a URL. Append the URL to the store. '''
    if urls >= history:
        iter = store.get_iter_first()
        store.remove( iter )
    else:
        urls += 1
    store.set( store.append(), 0, nick, 1, xchat.get_info('channel'), 2, match.group())

def grabbedEmail( nick, match ):
    ''' Handler for email addresses. Since the behaviour is the same
        as the URL handler we might as well just pass this on to it.
        The only difference is you can disable this one, so we check
        first.
        '''
    if scrape_email:
        grabbedURL( nick, match )

regexes = {
    re.compile('(ht|f)tps?://[^\s\>\]\)]+'):                 grabbedURL,
    re.compile('[\w\.\-\+]+@([0-9a-zA-Z\-]+\.)+[a-zA-Z]+'):  grabbedEmail,
    }

def grabURL( word, word_eol, user_data ):
    ''' Check a message for a URL. '''
    nick = word[0]
    message = word[1]
    for regex, handler in regexes.iteritems():
        for match in regex.finditer(message):
            handler(nick, match)
    return xchat.EAT_NONE

def unload( user_data ):
    ''' Destroy the window and stuff before we go away. '''
    window.destroy()
    print 'URL Scraper unloaded.'

# Track the number of URLs in the store
urls = 0

# Set up gconf stuff.
client = gconf.client_get_default()

# If our gconf keys don't exist...
if not client.dir_exists( '/apps/xchat/url_scraper' ):
    client.add_dir( '/apps/xchat/url_scraper', gconf.CLIENT_PRELOAD_NONE )
    history = 10
    scrape_email = False
    client.set_int( '/apps/xchat/url_scraper/history', history )
    client.set_bool( '/apps/xchat/url_scraper/scrape_email', scrape_email )
# If our gconf keys exist...
else:
    history = client.get_int( '/apps/xchat/url_scraper/history' )
    scrape_email = client.get_bool( '/apps/xchat/url_scraper/scrape_email' )

# Create the window.
window = gtk.Window( gtk.WINDOW_TOPLEVEL )
window.set_default_size( 300, 350 )
window.set_title( 'URL Scraper' )

scrolled = gtk.ScrolledWindow()
scrolled.set_policy( gtk.POLICY_AUTOMATIC, gtk.POLICY_AUTOMATIC )
store = gtk.ListStore( gobject.TYPE_STRING, gobject.TYPE_STRING, gobject.TYPE_STRING )

view = gtk.TreeView( store )
nick_col =  gtk.TreeViewColumn( 'Nick', gtk.CellRendererText(), text=0 )
nick_col.set_sort_column_id( 0 )
chan_col = gtk.TreeViewColumn( 'Channel', gtk.CellRendererText(), text=1)
chan_col.set_sort_column_id( 1 )
url_col =  gtk.TreeViewColumn( 'URL', gtk.CellRendererText(), text=2 )

view.append_column( nick_col )
view.append_column( chan_col )
view.append_column( url_col )

scrolled.add_with_viewport( view )
window.add( scrolled )
window.connect( 'destroy', close )
window.show_all()

xchat.hook_print( 'Channel Message', grabURL )
xchat.hook_unload( unload )

print 'URL Scraper loaded.'
### The End ###

File: trunk/xchat-gnome/plugins/url_scraper/urlscraper.c
#include <sys/types.h>
#include <regex.h>
#include <gtk/gtk.h>

#include "xchat-plugin.h"

#define VERSION "0.2"

static xchat_plugin *ph;	// Plugin handle.
static regex_t email;		// Regex that matches e-mail addresses.
static regex_t url;			// Regex that matches urls.
static int urls;			// Current total in the scraper.

static GtkWidget *window;
static GtkListStore *list_store;

static void make_window ()
{
	GtkScrolledWindow *scrolled;
	GtkWidget *treeview;
	GtkTreeViewColumn *nick_col, *chan_col, *url_col;
	GtkCellRendererText *nick_rend, *chan_rend, *url_rend;

	window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
	list_store = gtk_list_store_new (3, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING);

	scrolled = gtk_scrolled_window_new (NULL, NULL);
	treeview = gtk_tree_view_new ();

	nick_rend = gtk_cell_renderer_text_new ();
	nick_col = gtk_tree_view_column_new_with_attributes ("Nick", nick_rend, 0);

	chan_rend = gtk_cell_renderer_text_new ();
	chan_col = gtk_tree_view_column_new_with_attributes ("Channel", chan_rend, 0);

	url_rend = gtk_cell_renderer_text_new ();
	url_col = gtk_tree_view_column_new_with_attributes ("URL", url_rend, 0);

}

static int grabURL (char **word, char **userdata)
{
	return XCHAT_EAT_NONE;
}

int xchat_plugin_init (xchat_plugin *plugin_handle,
		char **plugin_name,
		char **plugin_desc,
		char **plugin_version,
		char *arg)
{
	ph = plugin_handle;

	*plugin_name = "URL Scraper";
	*plugin_desc = "Grabs URLs and puts them in a separate window for easy viewing.";
	*plugin_version = VERSION;

	regcomp (&email, "[\w\.\-\+]+@([0-9a-z\-]+\.)+[a-z]+", REG_ICASE);
	regcomp (&url, "(ht|f)tps?://[^\s\>\]\)]+", REG_ICASE);

	urls = 0;

	xchat_hook_print (ph, "Channel Message", XCHAT_PRI_NORM, grabURL, 0);

	xchat_print (ph, "URL Scrapler loaded.\n");

	return 1;
}

void xchat_plugin_get_info (char **plugin_name,
		char **plugin_desc,
		char **plugin_version)
{
	*plugin_name = "URL Scraper";
	*plugin_desc = "Grabs URLs and puts them in a separate window for easy viewing.";
	*plugin_version = VERSION;
}
